src/coreclr/vm/amd64/AsmHelpers.asm

```diff
@@ -698,5 +698,246 @@ NESTED_END OnCallCountThresholdReachedStub, _TEXT
 
 endif ; FEATURE_TIERED_COMPILATION
 
+ifdef FEATURE_GREENTHREADS
+
+extern AllocateMoreStackHelper:proc
+extern End_More_Thread_Bookeeping:proc
+
+;
+; Called from a function prolog (with argument registers holding current state)
+; RAX holds the size of the calling stack frame
+; Provides a guarantee of at least 2KB of stack
+; TODO Make another helper for cases where lots of stack are needed
+; The prologue of this function is duplicated for the ResumeSuspendedThreadHelper2 function
+;
+NESTED_ENTRY _more_stack, _TEXT
+        ;
+        ; Allocate space for XMM parameter registers and callee scratch area.
+        ;
+        alloc_stack     0e8h ;; TODO This is overly large
+
+        mov             r10, rbp
+
+        ;
+        ; Save integer parameter registers.
+        ;
+        save_reg_postrsp    rcx, 70h
+        save_reg_postrsp    rdx, 78h
+        save_reg_postrsp    r8,  80h
+        save_reg_postrsp    r9,  88h
+
+        save_reg_postrsp    rbp, 90h
+        save_reg_postrsp    rbx, 98h
+
+        save_xmm128_postrsp xmm0, 20h
+        save_xmm128_postrsp xmm1, 30h
+        save_xmm128_postrsp xmm2, 40h
+        save_xmm128_postrsp xmm3, 50h
+
+        set_frame       rbp, 0e0h
+    END_PROLOGUE
+        ; Store the old stack limit and base
+        mov             r11, gs:[10h]  
+        mov             [rbp - 020h], r11 ; Store the old stack limit
+        mov             r11, gs:[8h]   
+        mov             [rbp - 018h], r11 ; Store the old stack base
+
+        ;;;;;;;;;;;;;;;;; EVERYTHING ABOVE THIS MUST BE IDENTICAL between _more_stack and ResumeSuspendedThreadHelper2
+
+        ; Allocate new stack and initialize it
+        mov rcx, rax
+        mov rdx, rbp  ; Specify the address of where the rsp pointed at entering the function
+                      ; This provides both the place to copy the old arguments to the new space
+                      ; as well as where to stash the stack_limit and stack_base data
+        call AllocateMoreStackHelper
+    
+        ;
+        ; Restore parameter registers
+        ;
+        mov             rcx, [rsp + 70h]
+        mov             rdx, [rsp + 78h]
+        mov             r8,  [rsp + 80h]
+        mov             r9,  [rsp + 88h]
+        movdqa          xmm0, [rsp + 20h]
+        movdqa          xmm1, [rsp + 30h]
+        movdqa          xmm2, [rsp + 40h]
+        movdqa          xmm3, [rsp + 50h]
+
+        ; Swap to new stack
+        mov             rbx, rsp ; Stash old stack pointer into rbx (saved register)
+        mov             rsp, rax
```

Comment:
Probably ok for now, but wondering if it would be feasible to use rsp for the virtual stack pointer at all, as I figure the OS could interrupt the thread and end up using the virtual stack while it's close to the limit.

Reply:
Agreed. We've had an interesting idea around this. At the moment I leave a fair bit of space as a "red-zone" but I'm not convinced it is a good idea long-term. Send me a message on Teams to discuss whenever you'd like to.